ENH: add machinery to support Array API

tupui opened the pull request:
### Reference issue
RFC: https://github.com/scipy/scipy/issues/18286
Initial PR with some discussions: https://github.com/tupui/scipy/pull/24

### What does this implement/fix?

This PR adds the first necessary bricks to support the [Array API](https://data-apis.org/array-api/latest/index.html). And it demonstrates how to use these to add support to a module, here `scipy.cluster` which is fully covered.

#### The changes

It introduces the following.

Two global variables:

- `SCIPY_ARRAY_API`: either a bool, `"all"`, or a list of supported backend. For now these are `"cupy", "pytorch", "numpy", "numpy.array_api"`. The global variable can be adjusted dynamically. This is an optional flag. By default the backend is `numpy` and it only affects functions you modified with the helpers bellow.
- `SCIPY_DEVICE`: can be `"cpu", "gpu", "mps"` (something supported by the backend)

Three helpers:

- `array_namespace`: detect the namespace based on input arrays and do some input validation (like refusing to work with masked Array, please see the RFC.)
- `as_xparray`: a drop-in replacement for `np.asarray` with additional features like `copy, check_finite`.
- `as_xparray_namespace`: does both, detect the namespace and convert arrays in one step.

Some testing fixture:

- `array_api_compatible`: use a parametrisation to run a test on multiple Array backend.
- `skip_if_array_api`: don't run a test if `SCIPY_ARRAY_API` is on.
- `skip_if_array_api_gpu`: don't run a test if GPU is involved (also applies to PyTorch's MPS mode.)

A new runtime dependency is needed, [array-api-compat](https://github.com/data-apis/array-api-compat). It could easily be vendored if we want, but for now it seems better to just pull it from PyPi. In the end, we would like to push as much as possible all of our "tricks" there so that we do the same things across all libraries.

And finally, there is a new GitHub action workflow which runs `pytorch-cpu`.

#### How to add support for a new function?

I will show an example. Let's say we have this function (yes I am making it up now and is completely random, so don't run it 😅):

```python
def toto(a, b):
    a = np.asarray(a)
    b = np.asarray(b, copy=True)

    c = np.sum(a) - np.prod(b)

    # this is some C or Cython call
    d = cdist(c)

    return d
```

You would convert this like so:

```python
def toto(a, b):
    xp = array_namespace(a, b)
    a = xp.asarray(a)
    b = as_xparray(b, copy=True)

    c = xp.sum(a) - xp.prod(b)

    # this is some C or Cython call
    c = np.asarray(c)
    d = cdist(c)
    d = xp.asarray(d)

    return d
```

The key is that, going to C or Cython requires to go back to NumPy. Meaning that it will raise if you are on a GPU as the conversion would be silent otherwise which is not good. For CPU interactions, it should be a transparent operation.


#### How to test?

Just use the decorator:

```python
@array_api_compatible
def test_toto(self, xp):
    a = xp.asarray([1, 2, 3])
    b = xp.asarray([0, 2, 5])
    toto(a, b)
```

Then you can use `dev.py` with he new option `-b` or `--array-api-backend`:

```
python dev.py test -b numpy -b pytorch -s cluster
```

This automatically set `SCIPY_ARRAY_API` appropriately. And if you want to test a different device, set manually `SCIPY_DEVICE`. e.g. `SCIPY_DEVICE=mps python dev.py ...`

### Additional information

Thank you to everyone who helped put this together 🚀 Here are some additional resources:

Quick started from https://github.com/scipy/scipy/pull/15395 and sklearn's https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/_array_api.py

Array API in sklearn: https://github.com/scikit-learn/scikit-learn/issues/22352

sklearn implentation from PR https://github.com/scikit-learn/scikit-learn/pull/22554 and https://github.com/scikit-learn/scikit-learn/pull/25956

rgommers commented:
Thanks @tupui, great to see this getting into shape!

CI isn't too happy, it looks like a couple of assumptions about the numpy version and use of the compat layer outside of the experimental flag snuck in:
```
ModuleNotFoundError: No module named 'array_api_compat'
```
and
```
ModuleNotFoundError: No module named 'numpy.array_api'
```

rgommers commented on scipy/cluster/tests/test_vq.py:
```diff
@@ -270,37 +323,44 @@ def test_krandinit(self):
-            init_cov = np.cov(init, rowvar=0)
+            init = _krandinit(data, k, rng, xp)
+            orig_cov = xp.cov(data.T)
+            init_cov = xp.cov(init.T)
```
Just a note that this is correct: I forgot about `rowvar`, that's pretty ugly.

rgommers commented on scipy/cluster/tests/test_vq.py:
```diff
@@ -270,37 +323,44 @@ def test_krandinit(self):
-        x = np.array([1, 2, 3, 4, 10], dtype=float)
-        res = kmeans(x, 1, thresh=1e16)
-        assert_allclose(res[0], np.array([4.]))
+        x = xp.asarray([1, 2, 3, 4, 10], dtype=float)
```
minor: while we're at it, let's clean up `dtype=float`? Use `xp.float64` instead.

rgommers commented on scipy/cluster/vq.py:
```diff
@@ -530,28 +544,31 @@ def _krandinit(data, k, rng):
 
     if data.ndim == 1:
-        cov = np.cov(data)
+        cov = xp.cov(data)
```
`cov` is not an array API standard function, so this shouldn't work. I guess all tested libraries have it under the same name. It looks like only plain `cov` with no keyword args is used in a bunch of places, which can fairly easily be implemented with standard primitives like `sum`, `mean` and `dot`.

h-vetinari commented on scipy/_lib/_array_api.py:
```diff
@@ -0,0 +1,239 @@
+    return array_api_compat.array_namespace(*arrays)
+
+
+def as_xparray(
+    array, dtype=None, order=None, copy=None, *, xp=None, check_finite=True
+):
+    """Drop-in replacement for `np.asarray`.
```
Where does the name `as_xparray` come from? I couldn't find it in https://github.com/scipy/scipy/issues/18286 nor on the array-api [docs](https://data-apis.org/array-api/latest/search.html?q=as_xparray).

tupui replied:
The part around `copy, order` is from sklearn, `check_finite` is to cover our needs.

The function itself is not supposed to be in the standard, it's a helper for us. Basically it's `xp.asarray` but with extra things that we need or rely on for now.

h-vetinari replied:
OK thanks.

> Basically it's `xp.asarray` but with extra things that we need or rely on for now.

Do we have an idea what might change (re: "for now") in the future?

I'm not too keen on the name `as_xparray` TBH, as it mixes separation by `_` with lack of separation between words. Also, the `xp` is not self-explanatory, except perhaps where it's meant for 1:1 replacement of `np`. I recognize that this is what the array API requires (though I couldn't find what it stands for or a rationale[^1]), but do we _need_ to name our methods after it?

[^1]: I looked on the array-api docs, repo & conformance test suite. It seems to just appear [here](https://data-apis.org/array-api/latest/purpose_and_scope.html#how-to-adopt-this-api).

I apologise if this has been discussed/decided already, I'd be happy to read up on it (I looked for it in the issues you linked too). If not, I think there's room for a better name. Conservatively just `as_xp_array`, but also `asarray_xp`, or even `asarray_api` (which contracts both `np.asarray` as well as `array-api`). It's a pity that `scipy.asarray` is already taken...

tupui replied:
> Do we have an idea what might change (re: "for now") in the future?

I am really not sure here. Ideally we would want to remove as much specification as possible to get closer to the standard Array API. This way we would not need this anymore as our code would just be compliant by default.

> I'm not too keen on the name as_xparray

Happy to revisit the name 😃 I don't have a strong preference on my side. I kind of like your proposal of `asarray_xp`.

ev-br replied:
TBH it feels natural to fold this function into `xp.asarray`. I assume there was a prior discussion somewhere so that `array_api.asarray` does not get these extra keywords?

tupui commented on scipy/cluster/hierarchy.py:
```diff
@@ -2585,15 +2577,18 @@ def fcluster(Z, t, criterion='inconsistent', depth=2, R=None, monocrit=None):
             # Since the C code does not support striding using strides.
             # The dimensions are used instead.
             [R] = _copy_arrays_if_base_present([R])
+        R = np.asarray(R)
```
Yeah I can indent that one 👍

tupui commented on scipy/cluster/tests/test_vq.py:
```diff
@@ -71,37 +75,53 @@
-            assert_allclose(whiten(obs), desired, rtol=1e-5)
+            if "cupy" in xp.__name__:
+                import cupy as cp
+                cp.testing.assert_allclose(whiten(obs), desired, rtol=1e-5)
```
Yeah this would probably also help to have that for Pytorch. So makes sense to have a common function. But I suppose it would make sense to propose that to go in `array_api_compat`?

tupui commented on scipy/_lib/_array_api.py:
```diff
@@ -0,0 +1,239 @@
+    return np.asarray(array)
+
+
+def isdtype(dtype, kind, *, xp):
```
From sklearn

tupui commented on scipy/_lib/_array_api.py:
```diff
@@ -0,0 +1,239 @@
+    return *arrays, xp
+
+
+def to_numpy(array, xp):
```
From sklearn

tupui commented:
> CI isn't too happy, it looks like a couple of assumptions about the numpy version and use of the compat layer outside of the experimental flag snuck in:

Ark yes, I need to guard against that.

> * Some of the key parts of this PR were taken over from scikit-learn, where they were already thoroughly discussed and tested (e.g., dealing with the `order=` keyword). @tupui how about you mark those parts with comments in the PR? I think that will help the review, focusing on the new and SciPy-specific parts.

I added some comments.

> * Testing changes look quite good, very regular and easy to understand changes. The one thing that seemed not fully consistent is `@skip_if_array_api_gpu` usage - I'm guessing that testing on GPU needs re-running? Should I have a go at that?

Yes please :) 

> * There's a couple of places where functions are used that NumPy, CuPy and PyTorch all have but the array API standard does not (e.g., `cov`, `searchsorted`, `allclose`). So that works for all tested libraries, but shouldn't work for `numpy.array_api` I'd think. This can be dealt with as a follow-up issue perhaps; we can also do what scikit-learn did and document that PyTorch and CuPy were tested, other libraries should work but YMMV.

Yes +100, I need to run again on my side with `numpy.array_api`. I simply forgot as I was running all the time the same command for testing.

rgommers commented:
I see 81 failures with `python dev.py test -b numpy.array_api -s cluster`, and they almost all should be very easy to resolve (e.g., use `xp.min(x)` instead of `x.min()`). The `TypeError: unhashable type: 'ndarray'` (due to trying to use an array element in an `element in a_set` statement) overlap with errors I see with CuPy.

There's one other type of error visible with CuPy:
```
TypeError: Implicit conversion to a NumPy array is not allowed. Please use
`.get()` to construct a NumPy array explicitly.
```
That occurs quite a bit;  I assume the fix for all of them is to apply `@skip_if_array_api_gpu` to the tests.

asmeurer commented:
> TypeError: Implicit conversion to a NumPy array is not allowed. Please use
`.get()` to construct a NumPy array explicitly.

I think that error comes from trying to mix numpy and cupy. So it's probably worth investigating. Note that torch doesn't always give such an error:

```py
>>> torch.ones((2,)) + np.ones((2,))
tensor([2., 2.], dtype=torch.float64)
```

Also it can apparently sometimes come from just using cupy functions (unless this has been fixed) https://stackoverflow.com/questions/65008297/attempting-numpy-conversion-when-not-needed-in-cupy

asmeurer commented on scipy/cluster/hierarchy.py:
```diff
@@ -1052,29 +1027,33 @@ def linkage(y, method='single', metric='euclidean', optimal_ordering=False):
                              "to be Euclidean".format(method))
         if y.shape[0] == y.shape[1] and np.allclose(np.diag(y), 0):
-            if np.all(y >= 0) and np.allclose(y, y.T):
+            if xp.all(y >= 0) and xp.allclose(y, y.T):
```
Just to be clear, array_api_compat does, by design, include all functions from the wrapped module, even those not in the array API. So when `xp` is `numpy`, `xp.allclose` will work here. If you want to ensure portability across any potential array API namespace you need to test with `numpy.array_api`. 

It looks like allclose is in numpy, cupy, and torch. See also https://github.com/data-apis/array-api/issues/170

tupui replied:
yeah so that's one big question I suppose. If we are saying we are only supporting the strict Array API, hence yes test with `numpy.array_api` (which I forgot to do in my last pass), or we say we support a set of libraries (which happen to be the ones in `array_api_compat`, i.e. pushing some responsibility there).

rgommers replied:
> If we are saying we are only supporting the strict Array API, hence yes test with `numpy.array_api`

This. We should test with `numpy.array_api`. That was always the plan, and there will be needs - e.g., @tylerjereddy already indicated that there is interest to test a LANL-internal accelerator library. It's also the only thing that is well-defined; sticking to whatever happens to be the intersection of the NumPy, PyTorch and CuPy APIs that seems to pass our test suite is a bit of a moving target.

tylerjereddy reviewed (commented):
For `test_krandinit` failure with CuPy, I needed the obvious change on the assertion function, but also a bit of a tol bump. I suppose this might be a slight wrinkle re: the discussion I see above about using a wrapper that dispatches to i.e., lib-specific `assert_allclose()`.

```diff
--- a/scipy/cluster/tests/test_vq.py
+++ b/scipy/cluster/tests/test_vq.py
@@ -326,7 +326,11 @@ class TestKMean:
             init = _krandinit(data, k, rng, xp)
             orig_cov = xp.cov(data.T)
             init_cov = xp.cov(init.T)
-            assert_allclose(orig_cov, init_cov, atol=1e-2)
+            if "cupy" in xp.__name__:
+                import cupy as cp
+                cp.testing.assert_allclose(orig_cov, init_cov, atol=1.3e-2)
+            else:
+                assert_allclose(orig_cov, init_cov, atol=1e-2)
 
     @array_api_compatible
     def test_kmeans2_empty(self, xp):
```

When I looked at `test_kmeans2_high_dim` failure with CuPy, I found that a CuPy 0-D array was masquerading as an integer to `rng.standard_normal` in `_krandinit`. `_krandinit` documents `k` as `int` rather than array-like, though I suspect 0-D NumPy arrays have been allowed here for quite some time.

We can force it to an `int` like this:

```diff
--- a/scipy/cluster/vq.py
+++ b/scipy/cluster/vq.py
@@ -542,6 +542,8 @@ def _krandinit(data, k, rng, xp):
 
     """
     mu = data.mean(axis=0)
+    if not isinstance(k, int):
+        k = k.item()
 
     if data.ndim == 1:
         cov = xp.cov(data)
```

But is `item()` actually array API compatible? Quick search wasn't promising. Even with that patch, we end up at `cdist`, which I think we already agreed was a dead-end with CuPy for now in the other early-stage PR because of extra deps/complications. So, I guess that's basically a long way of saying skip `test_kmeans2_high_dim` on devices for the near future?

ev-br commented on scipy/_lib/_array_api.py:
```diff
@@ -0,0 +1,239 @@
+}
+
+
+def compliance_scipy(*arrays):
```
Nit: From the function name it's not immediately clear what this function is doing. Don't have an immediate suggestions though.

rgommers replied:
I also don't have a better suggestion. This is a private utility function only used in this file, so I'll resolve the comment - it doesn't matter much and can be tweaked later if desired.

ev-br commented on scipy/_lib/_array_api.py:
```diff
@@ -0,0 +1,239 @@
+    is NumPy based, otherwise `order` is just silently ignored.
+
+    The purpose of this helper is to make it possible to share code for data
+    container validation without memory copies for both downstream use cases.
```
Nit: which _both_ use cases?

ev-br commented on scipy/_lib/_array_api.py:
```diff
@@ -0,0 +1,239 @@
+    """Convert `array` into a NumPy ndarray on the CPU.
+
+    ONLY FOR TESTING
+    """
```
Why an extra `xp` argument, is there something which cannot be done by usual `isinstance(array, torch.Tensor)` etc?

ev-br commented on scipy/cluster/hierarchy.py:
```diff
@@ -1052,29 +1027,33 @@ def linkage(y, method='single', metric='euclidean', optimal_ordering=False):
     n = int(distance.num_obs_y(y))
     method_code = _LINKAGE_METHODS[method]
 
+    y = np.asarray(y)
```
Worth adding a comment is this convert to numpy -> do computations -> convert back is the idiom for interfacing with custom extensions (is this the reason it's done here?)

tupui replied:
Yep this is what I wanted to say in the description of the PR. When going to Cython you need to convert to NumPy. And once you are done you can come back to your backend.

I wrote a context manager if we want to use that instead, maybe that could help with the task 😃

ev-br reviewed (commented):
This looks great overall!

Several minor comments below, plus several bigger picture questions:

- (relatively minor) `as_xparray` vs `xp.asarray` split is suboptimal. So if the main reason for `as_xparray` is to have the `order` keyword, can we avoid `as_xparray` and make users spell it roughly
```
a = xp.asarray(a)
a = ensure_contiguous(a, order="C")
```
etc

- Would be great to have a spelled-out plan/guidance for how to convert other modules. E.g. how to deal with functions not available in the array_api namespace, do we have a central place for them (in `scipy._lib`)? Also recommendations for things which are commonly used but need shims? `atleast_1d`, `concatenate` and so on. The goal here would be to avoid proliferation of small module-specific shims and ad hoc replacements.
- Would be great to have a centralized tracker issue or a project to track which modules/functions are converted and which are not.

tupui commented:
> * (relatively minor) `as_xparray` vs `xp.asarray` split is suboptimal. So if the main reason for `as_xparray` is to have the `order` keyword, can we avoid `as_xparray` and make users spell it roughly
> 
> ```
> a = xp.asarray(a)
> a = ensure_contiguous(a, order="C")
> ```

This is just for conveniency for us. But agreed might be better to try to stick as much as possible with ArrayAPI and call explicitly our special things instead of mixing.

> * Would be great to have a spelled-out plan/guidance for how to convert other modules. E.g. how to deal with functions not available in the array_api namespace, do we have a central place for them (in `scipy._lib`)? Also recommendations for things which are commonly used but need shims? `atleast_1d`, `concatenate` and so on. The goal here would be to avoid proliferation of small module-specific shims and ad hoc replacements.

Yes I would say all extra functions like `atleast_1d` should go in `scipy._lib._array_api` so everything is self contained.

As we move forward and will add functions there, we will see with sklearn and other if there are common things to upstream to `array_api_compat`. That's one of the goal of this library.

e.g `concatenate` is taken care of so that we can do `xp.concat` with `array_api_compat.numpy`.

Circling back to `atleast_1d`, I just added a helper for that `scipy._lib._array_api.atleast_nd` (still need to add a few tests, etc.)

> * Would be great to have a centralized tracker issue or a project to track which modules/functions are converted and which are not.

Agreed, we can do this once this gets in.

tupui commented:
Next week I will add properly the deps to `array-api-compat`... 😅 For the import error on `numpy.array_api`, I was thinking about updating the minimal NumPy to 1.22 (another PR), we would either probably do this for SciPy 1.12.

ev-br commented:
I've played a bit with this and here's one more "what's the bigger plan" type question. So the ultimate goal is to use everywhere

```
In [11]: import numpy.array_api as np

```

and only operate with `numpy.array_api.Array` objects not `numpy.ndarray` for the numpy backend? 

If that's the case, what's the plan / recommendation for when the transition is not complete (also for downstream) and both `array_api.Array` and `numpy.ndarrays` are present?

Naively it seems any non-converted code needs to use an extra `np.asarray` to consume an `array_api.Array` object, is this really the recommendation?

```
In [12]: x= np.concat((np.asarray([1]), np.asarray([1, 2]), np.asarray([3])))       # here `np` is `numpy.array_api`

In [13]: type(x)
Out[13]: numpy.array_api._array_object.Array

In [14]: import numpy

In [15]: numpy.asarray(x)          # back to a plain ndarray
Out[15]: array([1, 1, 2, 3])
```

tupui commented:
@ev-br right, the long term goal is to write Array API compatible code. Not restricted to `numpy.array_api`.

The transition plan, for now, is to use `array_api_compat.numpy`. This wrapper should allow us to slowly migrate. If you look at the code right now, if the global flag is off, it would return `xp/np = array_api_compat.numpy` so when you are converting a function, it should still pass with plain NumPy and could break with a pure Array API array.

Did I answer your question fully??

tupui commented:
#18712 is needed to fix the failing tests about missing `numpy.array_api`.

For the rest, I now went ahead with adding `array_api_compat` as a submodule. The `meson.build` is probably adjustable 😅 (and yes, I add a skip for MyPy on this code...)

rgommers commented:
Changes to build config files and the git submodule LGTM; CI failures look real but don't seem too tricky to resolve.

ev-br commented:
> see https://github.com/data-apis/array-api-compat#difference-between-array_api_compat-and-numpyarray_api

@asmeurer thanks! In view of these differences, what is the point of using `np.array_api` at all? IOW, what are drawbacks of using `array_api_compat` only from the onset?

tupui commented:
> @asmeurer thanks! In view of these differences, what is the point of using `np.array_api` at all? IOW, what are drawbacks of using `array_api_compat` only from the onset?

The point of using `numpy.array_api` is to have a strict Array API compatible array. It's a testing conveniency not really meant to be used besides that. If you test your library with passing `numpy.array_api` arrays, it's a guarantee that you comply with the standard.

We use `array_api_compat.xp` because we (all libraries) are transitioning. NumPy, Cupy, etc. don't comply yet with the standard so we need that.

One drawback of using this compatibility layer is that one could still use non-standard methods without realising it. And this is why testing with a strict array is important (here we use `numpy.array_api`.)

tupui commented:
For these

```
FAILED scipy/cluster/tests/test_vq.py::TestWhiten::test_whiten[numpy] - TypeError: matrix.std() got an unexpected keyword argument 'keepdims'
FAILED scipy/cluster/tests/test_vq.py::TestWhiten::test_whiten_zero_std[numpy] - TypeError: matrix.std() got an unexpected keyword argument 'keepdims'
```

This is because in some test we are trying to use `np.matrix`. But this is incompatible with our usage of `array_api_compat.numpy`.

Here I just propose to remove the tests with `np.matrix`... If in the end we don't want to support that anymore, I think that's ok to stop testing with `np.matrix` arrays.

rgommers commented:
> Here I just propose to remove the tests with `np.matrix`...

That sounds like a good idea to me.

tupui commented:
All tests are fine locally. For the new CI job, I don't get the error. It's failing right after pip finishes to install deps and the message is cryptic. I will try to use Python 3.10, then it would be exactly the same as our other linux job.

Other CI jobs are blocked by the update to NumPy 1.22.4 (#18712)

tupui commented on .github/workflows/array_api.yml:
```diff
@@ -0,0 +1,86 @@
+    runs-on: ubuntu-22.04
+    strategy:
+      matrix:
+        python-version: ['3.11']
```
```suggestion
        python-version: ['3.10']
```

tupui commented on pytest.ini:
```diff
@@ -1,5 +1,5 @@
 [pytest]
-addopts = -l
+addopts = -l --ignore=scipy/_lib/array_api_compat
```
Still needed when running tests with `dev.py` as the test discovery will get there.

tupui commented on tools/check_installation.py:
```diff
@@ -45,6 +53,9 @@ def main(install_dir):
+        if test_file in exception_list_test_files:
+            continue
+
         if not test_file in installed_test_files.keys():
```
For the linter. Not linked to this PR.
```suggestion
        if test_file not in installed_test_files.keys():
```

tupui commented:
ok so CI is almost happy ("normal" failures). Now the issue is getting the windows job with `setup.py` working.

tupui commented:
If the cleaning of setup.py can be done now, that would indeed be simpler 😅

For the import, I am not sure how to modify the meson.build to make it work like you propose.

rgommers commented:
> It shouldn't look like that I think. Instead, the actual package should get installed as `scipy._lib import array_api_compat` rather than `scipy._lib import array_api_compat.array_api_compat`.

This was a bit of a pain also with the `setup.py` build, so I did the simpler thing - just adding the missing files as data files. Let's see if that turns CI green.

rgommers commented on scipy/cluster/hierarchy.py:
```diff
@@ -1550,29 +1537,30 @@ def optimal_leaf_ordering(Z, y, metric='euclidean'):
-def _convert_to_bool(X):
-    if X.dtype != bool:
-        X = X.astype(bool)
-    if not X.flags.contiguous:
```
This was also too well hidden inside a `_convert_to_xxx` function name. I'm considering getting rid of these functions completely.

rgommers replied:
This wasn't even used - I'm working on cleaning this up now.

rgommers replied:
This helper function is gone in `main` now, so resolving this comment.

rgommers commented on scipy/_lib/_array_api.py:
```diff
@@ -0,0 +1,184 @@
+    return array
+
+
+def as_xparray_namespace(*arrays):
```
This function isn't used at all anymore - should it be removed here?

rgommers commented on scipy/cluster/hierarchy.py:
```diff
@@ -2575,40 +2571,48 @@ def fcluster(Z, t, criterion='inconsistent', depth=2, R=None, monocrit=None):
+    Z = as_xparray(Z, order='c', xp=xp)
     is_valid_linkage(Z, throw=True, name='Z')
 
+    t = xp.asarray(t)
```
This should not be needed, and the `t_ = xp.astype(...)` doesn't seem right. This is documented as a scalar, and that's what the Cython code expects. It's unclear to me why this code wouldn't work without these changes.

rgommers commented on scipy/cluster/hierarchy.py:
```diff
@@ -2694,18 +2698,20 @@ def fclusterdata(X, t, criterion='inconsistent',
 
-    if type(X) != np.ndarray or len(X.shape) != 2:
-        raise TypeError('The observation matrix X must be an n by m numpy '
+    if len(X.shape) != 2:
```
minor: while we're at it, `X.ndim` seems more idiomatic than `len(X.shape)`

rgommers commented on scipy/cluster/hierarchy.py:
```diff
@@ -3558,8 +3572,8 @@ def _dendrogram_calculate_info(Z, p, truncate_mode,
     # Actual indices of a and b
-    aa = int(Z[i - n, 0])
-    ab = int(Z[i - n, 1])
+    aa = xp.astype(Z[i - n, 0], xp.int64)
```
This pattern occurs a lot. `int` should work I'd think - what am I missing? 

I tried with a GPU array quickly, and no problem there:
```python
>>> import torch
>>> t = torch.tensor(3, device='cuda')
>>> int(t)
3
>>> type(int(t))
<class 'int'>
```

tupui replied:
So, I guess there was something else I had to fix because now I could remove that. I just needed to keep the only float one because I had these issues otherwise:


```
# X = float(X)

X          = array([1])
E   DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)

y          = tensor(1.)
ValueError: `y` must be 1 or 2 dimensional.

X          = Array([1], dtype=int64)
TypeError: float is only allowed on arrays with 0 dimensions
```

tupui replied:
So, I guess there was something else I had to fix because now I could remove that. I just needed to keep the only float one because I had these issues otherwise:


```
# X = float(X)

X          = array([1])
E   DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)

y          = tensor(1.)
ValueError: `y` must be 1 or 2 dimensional.

X          = Array([1], dtype=int64)
TypeError: float is only allowed on arrays with 0 dimensions
```

rgommers replied:
Great - this looks good now.

rgommers commented on scipy/cluster/vq.py:
```diff
@@ -447,40 +461,42 @@ def kmeans(obs, k_or_guess, iter=20, thresh=1e-5, check_finite=True,
-    if not np.isscalar(k_or_guess):
-        guess = _asarray_validated(k_or_guess, check_finite=check_finite)
-        if guess.size < 1:
+    if size(guess) > 1:
```
This should be `!= 1`, otherwise the check on the next line will never work.

rgommers commented on scipy/cluster/vq.py:
```diff
@@ -447,40 +461,42 @@ def kmeans(obs, k_or_guess, iter=20, thresh=1e-5, check_finite=True,
             raise ValueError("Asked for 0 clusters. Initial book was %s" %
                              guess)
-        return _kmeans(obs, guess, thresh=thresh)
+        elif size(guess) > 1:
```
This `elif` seems unnecessary, it's always true once you arrive here

rgommers commented on scipy/cluster/vq.py:
```diff
@@ -578,18 +598,25 @@ def _kpp(data, k, rng):
+    # k should be an integer, NOT a NumPy
+    # scalar array thing...
+    if not isinstance(k, int):
+        k = xp.astype(k, xp.int64)
```
why not leave the above 4 lines out, and below use `init = xp.empty((int(k), dims))`?

rgommers reviewed (commented):
This is looking pretty good. After going through all the `hierarchy.py` and `vq.py` changes again, I think I like the patterns for asarray-like functions. There's really only two:

(1) Input array validation. Uses:
```python
xp = array_namespace(arr)  # where `arr` is the first input array
# Do this for each input array, it applies all the validation steps (reject matrix, etc.)
# as well as the conversion to a numpy array if it's a sequence,
# or preserve the non-numpy array type:
arr = as_xparray(arr)
```

(2) When hitting compiled code:
```python
x = np.asarray(x)  # convert to numpy right before compiled call(s)
y = _call_compiled_code(x)
y = xp.asarray(y)  # convert back to original array type
```

There's probably some room to reduce the to-from conversions for (2) in case a public function calls other public functions. It's not that expensive though, since it's all zero-copy. So this is better left for later, since this PR is already very large and restructuring code would make it very difficult to review.

asmeurer commented:
> This is because in some test we are trying to use np.matrix. But this is incompatible with our usage of array_api_compat.numpy.

Interesting

```py
>>> import array_api_compat.numpy as np
>>> a = np.matrix([[1, 2], [3, 4]])
>>> np.std(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/Documents/array-api-compat/./array_api_compat/_internal.py", line 27, in wrapped_f
    return f(*args, xp=xp, **kwargs)
  File "/Users/aaronmeurer/Documents/array-api-compat/./array_api_compat/common/_aliases.py", line 249, in std
    return xp.std(x, axis=axis, ddof=correction, keepdims=keepdims, **kwargs)
  File "<__array_function__ internals>", line 180, in std
  File "/Users/aaronmeurer/anaconda3/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 3613, in std
    return std(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)
TypeError: std() got an unexpected keyword argument 'keepdims'
```

`np.std(x)` dispatches to `x.std()` and `keepdims` is not supported in `matrix.std()`. We could fix this in the compat layer. The code assumes that keyword arguments can always be passed through, but we could instead only pass through `keepdims` when it has a non-default value. I guess matrix isn't super important but in general I don't want to break support for numpy.ndarray subclasses if it isn't necessary.

tupui commented on scipy/_lib/_array_api.py:
```diff
@@ -0,0 +1,184 @@
+    """
+    if xp is None:
+        xp = array_namespace(array)
+    if xp.__name__ in {"numpy", "array_api_compat.numpy", "numpy.array_api"}:
```
AFAIU, see the comment bellow, the idea is to use normal NumPy and then convert back to `numpy.array_api`.

(This is from sklearn.)

tupui replied:
AFAIU, see the comment bellow, the idea is to use normal NumPy and then convert back to `numpy.array_api`.

(This is from sklearn.)

rgommers replied:
Ah, I see what is going on there. That seems like the wrong thing to do to me. The whole point of using `numpy.array_api` is to test with a strictly-compliant implementation of the standard, so avoiding that by converting to regular `numpy` kinda defeats the purpose. I think we want `numpy.array_api.Array` instances to take the non-numpy-specific route here.

There's a second issue in this line I just noticed: the `array_api_compat` name is incorrect because the module is now vendored, it should be `"scipy._lib.array_api_compat.array_api_compat.numpy"`.

tupui replied:
It's aliased `import scipy._lib.array_api_compat.array_api_compat as array_api_compat`

rgommers replied:
Yes, but you're comparing exact module names; `__name__` doesn't care about what the import alias is. You can `print(xp)` to see it.

I think this is the change to apply:

```diff
diff --git a/scipy/_lib/_array_api.py b/scipy/_lib/_array_api.py
index d7d184731..fbc39bd21 100644
--- a/scipy/_lib/_array_api.py
+++ b/scipy/_lib/_array_api.py
@@ -109,7 +109,7 @@ def as_xparray(
     """
     if xp is None:
         xp = array_namespace(array)
-    if xp.__name__ in {"numpy", "array_api_compat.numpy", "numpy.array_api"}:
+    if xp.__name__ in {"numpy", "scipy._lib.array_api_compat.array_api_compat.numpy"}:
         # Use NumPy API to support order
         if copy is True:
             array = np.array(array, order=order, dtype=dtype)
```

tupui commented on scipy/_lib/_array_api.py:
```diff
@@ -0,0 +1,184 @@
+
+
+def as_xparray(
+    array, dtype=None, order=None, copy=None, *, xp=None, check_finite=True
```
Oops just a mistake from my side.

tupui commented:
So, I think I addressed the comments and hopefully this will come back green too (locally all good.) I also added a guide to migrate. LMK what you think.

j-bowhay commented on LICENSES_bundled.txt:
```diff
@@ -256,3 +256,8 @@ Files: scipy/stats/_rcont/[logfactorial.h,logfactorial.c]
+Name: array-api-compat
+Files: scipy/_lib/array-api-compat/*
+License: MIT
+  For details, see scipy/optimize/_highs/LICENCE
```
Think this is pointing to the wrong file

tupui replied:
ark, good catch thanks!
```suggestion
  For details, see scipy/_lib/array-api-compat/LICENCE
```

tupui replied:
ark, good catch thanks!

tupui commented:
Oh waou CI is all green! Are we ready for another round on the mailing list?

rgommers commented:
@tupui I think we should remove this marker:
```python
skip_if_array_api = pytest.mark.skipif(
    SCIPY_ARRAY_API,
    reason="do not run with Array API on",
)
```
I was quite puzzled why, when setting the env var, less tests were run with `test -s cluster -b numpy` than if the env var was off. Those tests should all pass anyway, and I don't see a good reason to disable them. Or am I missing something?

rgommers commented:
@tupui I updated the docs, and made sure the tests pass with `cupy` and `numpy.array_api`. It looks ready to me now. 

I also had a quick play with calling CuPy's implementation of `cdist`/`pdist`, because those functions are used heavily in `cluster` and CuPy has matching implementations. That makes some more things work, e.g.:

```diff
diff --git a/scipy/spatial/distance.py b/scipy/spatial/distance.py
index 00c604410..7b177cb77 100644
--- a/scipy/spatial/distance.py
+++ b/scipy/spatial/distance.py
@@ -2972,6 +2972,11 @@ def cdist(XA, XB, metric='euclidean', *, out=None, **kwargs):
     # between all pairs of vectors in XA and XB using the distance metric 'abc'
     # but with a more succinct, verifiable, but less efficient implementation.
 
+    from scipy._lib.array_api_compat.array_api_compat.common._helpers import _is_cupy_array
+    if _is_cupy_array(XA):
+        import cupyx.scipy.spatial.distance as cupy_distance
+        return cupy_distance.cdist(XA, XB, metric=metric, out=out, **kwargs)
+
     XA = np.asarray(XA)
     XB = np.asarray(XB)
```
I didn't commit that; the scope of this PR should remain limited to `cluster`. But it's nice to know that it's possible, and could be a useful follow-up.

tupui commented:
Great thank you for the help! (Also I see you had my int issue and from the commit message I understand why I did not have it anymore as I updated my local NumPy.)

I will send an email to the list now then and hopefully we can more forward then!

rgommers commented:
It would be very useful if someone not directly involved with this PR would read/review the page added under _SciPy API Development Guide_ (latest rendered version: [here](https://output.circle-artifacts.com/output/job/ba6f1506-d256-476a-b7d0-a4a89a7c25f3/artifacts/0/html/dev/api-dev/array_api.html)). Maybe @mdhaber, @ev-br, or @WarrenWeckesser?

ev-br commented:
+1 from me. I think the current state sets reasonable patterns, the documentation is nice and enough to get going. We would surely find a thing or two to adjust based on further usage--- so it's IMO best to get this in and get this further usage by converting other subpackages.

Full disclosure:

1. I'm on record for a preference for iterations.
2. I had several offline discussions about this PR with Ralf and Pamphile.

lorentzenchr commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -0,0 +1,253 @@
+This guide describes how to **use** and **add support for** the
+`Python Array API standard <https://data-apis.org/array-api/latest/index.html>`_.
+This standard allows users to use any Array API compatible array library
+without having to do anything.
```
```suggestion
with scipy out of the box.
```

lorentzenchr commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -0,0 +1,253 @@
+
+.. code:: bash
+
+   export SCIPY_ARRAY_API=1
```
What‘s the reason to use an env variable? Why not global config option? A short comment would be nice.

tupui replied:
We have both. It's the same behaviour as in Scikit-Learn.

rgommers replied:
Not quite - scikit-learn provides a context manager. And our global config is explicitly private; we're aiming for it to go away as soon as reasonably possible, it's only a tool to unblock development. Scikit-learn has public global config.

So the basic reason is that scikit-learn already had it, SciPy doesn't (and we'd prefer to keep it that way).

rgommers replied:
I am adding this note in the docs here to address the review comment:

*Note that this environment variable is meant to be temporary, as a way to make incremental changes and merge them into ``main`` without affecting backwards compatibility immediately. We do not intend to keep this environment variable around long-term.*

lorentzenchr commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -0,0 +1,253 @@
+for other libraries like JAX is expected to be added in the future).
+
+When the environment variable isn't set and hence Array API support in SciPy is
+disabled, we still use the "augmented" version of the NumPy namespace, which is
```
What does that mean? Will it accept a pytorch tensor? Will it convert the tensor to a numpy array?

tupui replied:
If the flag is off, it's as is we were not doing anything (so no, it only accepts NumPy arrays). Except that we don't use the plain NumPy namespace but we use `array_api_compat.numpy`. This is because as we update the code to the standard, we need functions from the standard like `xp.concat` which is not present in plain NumPy.

lorentzenchr commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -0,0 +1,253 @@
+  of non standard features. In the end we would want to upstream our needs to
+  the compatibility library.
+
+To add support to a Python-level SciPy function, what you have to change is:
```
```suggestion
To add support to a Python-level SciPy function, i.e. no compiled code involved, what you have to change is:
```

rgommers replied:
This seems incorrect, given point (3) below. It's any function that has a Python `def ...` entry point. I'll rephrase.

lorentzenchr commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -0,0 +1,253 @@
+
+Input array validation uses the following pattern::
+
+   xp = array_namespace(arr)  # where `arr` is the first input array
```
What if a function has 2 array arguments and a user gives a list to the first and a pytorch tensor to the second?

tupui replied:
Only Array API compatible arrays are supported, meaning that if you pass a list (does not matter if you pass one or n arrays), it will get rejected.

The case of mixed type is also not supported and would raise. e.g. passing a CuPy array and a PyTorch tensor.

rgommers replied:
We can spell this out more explicitly in the guide perhaps, since it's been asked about multiple times. For folks familiar with PyTorch, CuPy etc. this makes perfect sense, but NumPy has long instilled an "anything goes" expectation into its users - and we're trying to get away from that, or at least not make it worse.

rgommers replied:
I am adding this to the docs:

*Note that if one input is a non-numpy array type, all array-like inputs have to
be of that type; trying to mix non-numpy arrays with lists, Python scalars or
other arbitrary Python objects will raise an exception. For NumPy arrays, those
types will continue to be accepted for backwards compatibility reasons.*

lorentzenchr commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -0,0 +1,253 @@
+
+  python dev.py test -b numpy -b pytorch -s cluster
+
+This automatically set ``SCIPY_ARRAY_API`` appropriately. And if wanted, to
```
```suggestion
This automatically sets ``SCIPY_ARRAY_API`` appropriately. And if wanted, to
```

lorentzenchr commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -0,0 +1,253 @@
+  python dev.py test -b numpy -b pytorch -s cluster
+
+This automatically set ``SCIPY_ARRAY_API`` appropriately. And if wanted, to
+test a different device, a second environment variable only used in the test
```
This sentence could be made easier to read.

lorentzenchr commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -0,0 +1,253 @@
+helped during the development phase:
+
+* Initial `PR <https://github.com/tupui/scipy/pull/24>`__ with some discussions
+* Quick started from this `PR <https://github.com/scipy/scipy/pull/15395>`__ and
```
```suggestion
* Quick start from this `PR <https://github.com/scipy/scipy/pull/15395>`__ and
```

lorentzenchr commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -0,0 +1,253 @@
+
+* Initial `PR <https://github.com/tupui/scipy/pull/24>`__ with some discussions
+* Quick started from this `PR <https://github.com/scipy/scipy/pull/15395>`__ and
+  take some inspiration from
```
```suggestion
  some inspiration taken from
```

lorentzenchr reviewed (commented):
Those are my comments/questions while reading the dev guide.

tupui reviewed (commented):
Thank you @lorentzenchr for your comments. I hope I addressed your questions.

mdhaber reviewed (approved):
Agreed with https://github.com/scipy/scipy/pull/18668#issuecomment-1620534658. I haven't reviewed this in detail (just skimming code and reading documentation), but I will only be able to form a solid opinion by trying it out, so let's get this in.

I'll mention a preliminary preference for helper functions to instead be worked into the `xp` namespace so we don't have a mix of e.g. `as_xparray` and `xp.asarrray`. But again, I think I'll be able to form a more complete opinion by trying it out.

ev-br commented:
> the xp namespace would have .random and .special sub-namespaces with functions that mirror the interface of numpy.random and scipy.special where possible

IIUC, this is not possible really. For one, backend's `random` implementations generate different random streams. 
The best idiom I can think of is to generate variates with numpy and wrap to the target namespace:
`xp.asarray(np.random.uniform(...))`  or `rng = np.default_rng(seed); variates = xp.asarray(rng.uniform(size=size))`.

rgommers commented:
> If it were up to me right now, I think the `xp` namespace would have `.random` and `.special` sub-namespaces with functions that mirror the interface of `numpy.random` and `scipy.special` where possible (mapping to the underlying functions of the backend as needed)

A couple of comments:
- The `xp` namespace shouldn't be extended, it should exactly match what's in the array API standard. Anything else is going to be confusing. We can have other things in another single namespace, e.g. under `scipy._lib`.
- `random` and `special` are quite different cases:
- `special` is more straightforward: CuPy and PyTorch have matching namespaces with a reasonable amount of coverage, [cupyx.scipy.special](https://docs.cupy.dev/en/stable/reference/scipy_special.html) and [torch.special](https://pytorch.org/docs/master/special.html). As does JAX: [jax.scipy.special](https://jax.readthedocs.io/en/latest/jax.scipy.html#module-jax.scipy.special). Special functions are low-level and almost always implemented in compiled code, so we do need matching functions for the input array type, and then we call those. If they're not available, we can go through NumPy on CPU and raise an exception on other devices.
- `random` is a different story.  As far as I know, only CuPy has an API that more or less matches the new NumPy API: https://docs.cupy.dev/en/stable/reference/random.html#new-random-generator-api - with a different set of bit generators, but still it has the right methods so such instances could be accepted. PyTorch only has something that roughly matches the old NumPy API. And JAX has a fundamentally different design, functional rather than class-based. 

> For one, backend's `random` implementations generate different random streams.

I think that'd in principle be fine for, for example, `scipy.stats` functions that accept generator instances. Only for tests where we match expected results to a hardcoded seed would it be a problem.

h-vetinari commented:
> I'll mention a preliminary preference for helper functions to instead be worked into the `xp` namespace so we don't have a mix of e.g. `as_xparray` and `xp.asarrray`.

Also not a fan of this divergence. There was a [thread](https://github.com/scipy/scipy/pull/18668#discussion_r1228844818) in this PR where people were mostly in favour of coming up with a better name, but I guess we can also defer that decision until after this PR. I just don't want it to ossify into the status quo. `as_xparray` vs. `xp.asarrray` contains at least three stumbling blocks for remembering and understanding the distinction, so I'd like to come up with a better name. I can open an issue if we defer this.

rgommers commented:
I was planning to open a tracking issue with follow-up actions. If you already want to open a separate issue for a better name for `as_xparray` @h-vetinari, I think that would be helpful. Choosing names is hard, good ideas welcome:)

rgommers commented:
There's (at least) one more update to make here: after gh-18858 gets merged, this PR needs to be updated for conflicts.

rgommers commented on scipy/cluster/hierarchy.py:
```diff
@@ -4082,8 +4139,9 @@ def leaders(Z, T):
-    T = np.asarray(T, order='C')
+    xp = array_namespace(Z, T)
+    Z = as_xparray(Z, order='C', dtype=xp.float64)
+    T = as_xparray(T, order='C')
```
one issue snuck in here, the `if type(T) != np.ndarray ` below is causing a test failure

rgommers reviewed (approved):
This is now in quite good shape, and with several thumbs-up's it's time to merge this. Thanks for the hard work @tupui, @tylerjereddy and everyone else who contributed!

I went through all comments again, and here are the needed follow-ups and loose ends:

- a better name for `as_xparray`
- a TODO comment in `hierarchy.py` for rewriting a fancy indexing expression (may need `xp.put` to be standardized, see [this comment](https://github.com/scipy/scipy/pull/18668#issuecomment-1594618749))
- there's a usage of `xp.cov` that needs changing (it's in numpy/cupy/pytorch, but not in the standard)
- consider removing the `skip_if_array_api` marker (see [this comment](https://github.com/scipy/scipy/pull/18668#issuecomment-1614945157))
- (after we drop `setup.py`) making the import of the vendored `array_api_compat` nicer (see [this comment](https://github.com/scipy/scipy/pull/18668#issuecomment-1607878468))

And then next steps:
- open a tracking issue for the above issues and another one for overall coverage (I'll do that today)
- work on other scipy submodules now that we have the patterns and CI/infra for this
- reduce the number of test skips

asmeurer commented:
> 
I was planning to open a tracking issue with follow-up actions. If you already want to open a separate issue for a better name for as_xparray @h-vetinari, I think that would be helpful. Choosing names is hard, good ideas welcome:)

I think we decided that this specific wrapper was doing things that weren't appropriate for general inclusion in the compat library (c.f. https://github.com/data-apis/array-api-compat/pull/42), but an option could be to monkeypatch `array_api_compat.asarray`, especially since you are vendoring it.

rgommers commented:
Thanks for the suggestion @asmeurer. I'll move your comment to the follow-up issue (gh-18866) now.

lucascolley commented on scipy/_lib/_array_api.py:
```diff
@@ -0,0 +1,138 @@
+    1. Check for the global switch: SCIPY_ARRAY_API. This can also be accessed
+       dynamically through ``_GLOBAL_CONFIG['SCIPY_ARRAY_API']``.
+    2. `compliance_scipy` raise exceptions on known-bad subclasses. See
+       it's definition for more details.
```
typo: ~it's~ its

mdhaber commented on scipy/cluster/hierarchy.py:
```diff
@@ -995,7 +1002,8 @@ def linkage(y, method='single', metric='euclidean', optimal_ordering=False):
     """
-    y = np.asarray(y, order='C', dtype=np.float64)
+    xp = array_namespace(y)
+    y = as_xparray(y, order='C', dtype=xp.float64, xp=xp)
```
I'm working on gh-21835 and this confused me, since `order` only works for NumPy arrays.

It looks like `order='C'` was added in https://github.com/scipy/scipy/commit/57e1ec714bbcbb2ce6a4bbc76362b5071e294890 to ensure that the data is contiguous before it goes into compiled code. But it seems like non-contiguous arrays from other libraries will get past this and not be forced to be contiguous when they are turned into NumPy arrays. 

It this still important? Or at this point, does our machinery recognize that the arrays are not contiguous and copy the data to be contiguous (and in the appropriate order) before passing it into compiled code?

Should there be an error when the array library does not support options passed into the `_asarray` function?
